cpp/src/strings/reverse.cu

```diff
@@ -0,0 +1,89 @@
+/*
+ * Copyright (c) 2022, NVIDIA CORPORATION.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include <cudf/column/column_device_view.cuh>
+#include <cudf/column/column_factories.hpp>
+#include <cudf/detail/nvtx/ranges.hpp>
+#include <cudf/strings/detail/utf8.hpp>
+#include <cudf/strings/reverse.hpp>
+#include <cudf/strings/string_view.cuh>
+#include <cudf/strings/strings_column_view.hpp>
+#include <cudf/utilities/default_stream.hpp>
+#include <cudf/utilities/error.hpp>
+
+#include <rmm/cuda_stream_view.hpp>
+#include <rmm/exec_policy.hpp>
+
+namespace cudf {
+namespace strings {
+namespace detail {
+namespace {
+
+/**
+ * @brief Reverse individual characters in each string
+ */
+struct reverse_characters_fn {
+  column_device_view const d_strings;
+  offset_type const* d_offsets;
+  char* d_chars;
+
+  __device__ void operator()(size_type idx)
+  {
+    if (d_strings.is_null(idx)) { return; }
+    auto const d_str = d_strings.element<string_view>(idx);
+    // pointer to the end of the output string
+    auto d_output = d_chars + d_offsets[idx] + d_str.size_bytes();
+    for (auto const chr : d_str) {          // iterate through each character;
+      d_output -= bytes_in_char_utf8(chr);  // position output;
+      from_char_utf8(chr, d_output);        // place character into output
+    }
+  }
+};
+
+}  // namespace
+
+std::unique_ptr<column> reverse(strings_column_view const& input,
+                                rmm::cuda_stream_view stream,
+                                rmm::mr::device_memory_resource* mr)
+{
+  if (input.is_empty()) { return make_empty_column(type_id::STRING); }
+
+  // copy the column; replace data in the chars column
+  auto result = std::make_unique<column>(input.parent(), stream, mr);
```

Comment:
Is this making a full copy of the string data and then replacing it in the functor?

Reply:
Yes, that was the intention of the comment. Should've been worded better?

Reply:
I see that the char child of the result column will be overwritten. Should we just copy the offsets over?

Reply:
This one step does everything we need. It builds the output column and normalizes the offsets if the input is sliced.